artiq/coredevice/shuttler.py

```diff
@@ -89,3 +124,140 @@ def __init__(self, dmgr, channel, core_device="core"):
     @kernel
     def trigger(self, trig_out):
         rtio_output(self.target_o, trig_out)
+
+
+RELAY_SPI_CONFIG = (0*spi.SPI_OFFLINE | 1*spi.SPI_END |
+                    0*spi.SPI_INPUT | 0*spi.SPI_CS_POLARITY |
+                    0*spi.SPI_CLK_POLARITY | 0*spi.SPI_CLK_PHASE |
+                    0*spi.SPI_LSB_FIRST | 0*spi.SPI_HALF_DUPLEX)
+
+SPI_CS_RELAY = 1 << 0
+SPI_CS_LED = 1 << 1
+SPI_DIV = 4
+
+ADC_SPI_CONFIG = (0*spi.SPI_OFFLINE | 0*spi.SPI_END |
+                  0*spi.SPI_INPUT | 0*spi.SPI_CS_POLARITY |
+                  1*spi.SPI_CLK_POLARITY | 1*spi.SPI_CLK_PHASE |
+                  0*spi.SPI_LSB_FIRST | 0*spi.SPI_HALF_DUPLEX)
+
+ADC_CS = 1
+ADC_SPI_DIV = 32
+
+
+class Relay:
+    kernel_invariant = {"core", "bus"}
+
+    def __init__(self, dmgr, spi_device, core_device="core"):
+        self.core = dmgr.get(core_device)
+        self.bus = dmgr.get(spi_device)
+    
+    @kernel
+    def init(self):
+        self.bus.set_config_mu(
+            RELAY_SPI_CONFIG, 16, SPI_DIV, SPI_CS_RELAY | SPI_CS_LED)
+
+    @kernel
+    def set_led(self, leds: TInt32):
+        self.bus.write(leds << 16)
+
+
+class ADC:
+    kernel_invariant = {"core", "bus"}
+
+    def __init__(self, dmgr, spi_device, core_device="core"):
+        self.core = dmgr.get(core_device)
+        self.bus = dmgr.get(spi_device)
+
+    @kernel
+    def read_id(self) -> TInt32:
+        self.bus.set_config_mu(
+            ADC_SPI_CONFIG | spi.SPI_END | spi.SPI_INPUT,
+            24, ADC_SPI_DIV, ADC_CS)
+        self.bus.write(0x47 << 24)
+        return (self.bus.read() & 0xFFFF)
+    
+    @kernel
+    def read_ch(self, channel: TInt32) -> TFloat:
+        # Always configure Profile 0
+        self.bus.set_config_mu(
+            ADC_SPI_CONFIG | spi.SPI_END, 24, ADC_SPI_DIV, ADC_CS)
+        self.bus.write(0x10 << 24 | (0x8000 | ((channel * 2 + 1) << 4)) << 8)
+
+        # Configure Setup 0
+        # Input buffer must be enabled to use REF pins correctly
+        self.bus.set_config_mu(
+            ADC_SPI_CONFIG | spi.SPI_END, 24, ADC_SPI_DIV, ADC_CS)
+        self.bus.write((0x20 << 24) | (0x1300 << 8))
+
+        # Trigger single conversion
+        self.bus.set_config_mu(
+            ADC_SPI_CONFIG, 24, ADC_SPI_DIV, ADC_CS)
+        self.bus.write((0x01 << 24) | (0x8010 << 8))
+
+        self.bus.set_config_mu(
+            ADC_SPI_CONFIG | spi.SPI_INPUT, 16, ADC_SPI_DIV, ADC_CS)
+        self.bus.write(0x40 << 24)
+        while self.bus.read() & 0x80:
+            delay(10*us)
+            self.bus.write(0x40 << 24)
```

Comment:
While polling the `RDY` flag is not wrong, I think I can just give an upperbound of the waiting time. Sampling rate is configurable in the ADC setups.